File: observability/trace_execve/trace_execve.py

# 引入库函数
from bcc import BPF
from bcc.utils import printb

# 加载eBPF代码
b = BPF(src_file="trace_execve.c")

print("%-6s %-16s %-3s %s" % ("PID", "COMM", "RET", "ARGS"))

# 定义性能事件打印函数
def print_event(cpu, data, size):
    # BCC自动根据"struct data_t"生成数据结构
    event = b["events"].event(data)
    printb(b"%-6d %-16s %-3d %-16s" % (event.pid, event.comm, event.retval, event.argv))

# 绑定性能事件映射和输出函数，并从映射中循环读取数据
 
# 1. Perf/Ring Buffer相对于其他种类map(被动轮询)来说，提供专用api，通知应用层事件就绪，减少cpu消耗，提高性能。
# 2. 采用共享内存，节省复制数据开销。
# 3. Perf/Ring Buffer支持传入可变长结构。
# 差异:
# 1. Perf Buffer每个CPU核心一个缓存区，不保证数据顺序(fork exec exit)，会对我们应用层消费数据造成影响。Ring Buffer多CPU共用一个缓存区且内部实现了自旋锁，保证数据顺序。
# 2. Perf Buffer有着两次数据拷贝动作，当空间不足时，效率低下。 Ring Buffer采用先申请内存，再操作形式，提高效率。
# 3. perfbuf 的 buffer size 是在用户态定义的，而 ringbuf 的 size 是在 bpf 程序中预定义的。
# 4. max_entries 的语义， perfbuf 是 buffer 数量(社区推荐设置为cpu个数)，ringbuf 中是单个 buffer 的 size。
# 5. Ring Buffer性能强于Perf Buffer。参考patch 【ringbuf perfbuf 性能对比】
b["events"].open_perf_buffer(print_event)
while True:
    try:
        b.perf_buffer_poll()
    except KeyboardInterrupt:
        exit()

File: observability/trace_execve/trace_execve.c
// ls /sys/kernel/debug/tracing/events 查看所有挂载点
#include <uapi/linux/ptrace.h>
#include <linux/sched.h>
#include <linux/fs.h>

#define ARGSIZE 64
#define TOTAL_MAX_ARGS 5
#define FULL_MAX_ARGS_ARR (TOTAL_MAX_ARGS * ARGSIZE)
#define LAST_ARG (FULL_MAX_ARGS_ARR - ARGSIZE)

// perf缓存
BPF_PERF_OUTPUT(events);

// ebpf map, key为进程id
struct data_t {
	u32 pid;
	char comm[TASK_COMM_LEN];
	int retval;
	unsigned int args_size;
	char argv[FULL_MAX_ARGS_ARR];
};
BPF_HASH(tasks, u32, struct data_t);

// 读取用户空间字符串
static int __bpf_read_arg_str(struct data_t *data, const char *ptr)
{
	if (data->args_size > LAST_ARG) {
		return -1;
	}
	int ret = bpf_probe_read_user_str(&data->argv[data->args_size], ARGSIZE,
					  (void *)ptr);
	if (ret > ARGSIZE || ret < 0) {
		return -1;
	}
	data->args_size += (ret - 1);
	return 0;
}

// 进入sys_execve
TRACEPOINT_PROBE(syscalls, sys_enter_execve)
{
	unsigned int ret = 0;
	const char **argv = (const char **)(args->argv);
	struct data_t data = { };
	u32 pid = bpf_get_current_pid_tgid();
	data.pid = pid;
	bpf_get_current_comm(&data.comm, sizeof(data.comm));
	// 读取文件名
	if (__bpf_read_arg_str(&data, (const char *)argv[0]) < 0) {
		goto out;
	}
	// 获取其他参数
	// ebpf的循环需要使用宏展开
#pragma unroll
	for (int i = 1; i < TOTAL_MAX_ARGS; i++) {
		if (__bpf_read_arg_str(&data, (const char *)argv[i]) < 0) {
			goto out;
		}
	}
 out:
	// 存储到map结构中
	tasks.update(&pid, &data);
	return 0;
}

// 字符串比较
static int cmp_str(const char* s1, const char* s2) {
	if (!s1 || !s2) {
		return -1;
	}
	if ((*s1 != '\0') && (*s2 == *s2)) {
		++s1;
		++s2;
	}
	if (*s1 > *s2) {
		return 1;
	} else if (*s1 < *s2) {
		return -1;
	} else {
		return 0;
	}
}

// 退出sys_execve
TRACEPOINT_PROBE(syscalls, sys_exit_execve)
{
	u32 pid = bpf_get_current_pid_tgid();
	struct data_t *data = tasks.lookup(&pid);
	if (data != NULL) {
		data->retval = args->ret;
		if (!cmp_str("ls", data->comm)) {
			events.perf_submit(args, data, sizeof(struct data_t));
		}
		tasks.delete(&pid);
	}
	return 0;
}